solve_trig1 now able to convert hyperbolic-trig to exp

ishanaj opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234". See
https://github.com/blog/1506-closing-issues-via-pull-requests . Please also
write a comment on that issue linking back to this pull request once it is
open. -->
Fixes #9606 
Closes #9531

#### Brief description of what is fixed or changed
solveset returns a `ConditionSet` for hyperbolic-trig equations, as it is not able to convert it to `exp` form.
`solveset` would now return:
```
>>>solveset(sinh(x), x)
Union(ImageSet(Lambda(_n, 2*_n*pi), S.Integers), ImageSet(Lambda(_n, 2*_n*pi + pi), S.Integers))
```

ping @smichr @ylemkimon @Yathartha22 
#### Other comments

#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* solvers
   * Solving of hyperbolic equations improved in solveset.
<!-- END RELEASE NOTES -->

ishanaj committed:
Merge pull request #2 from sympy/master

update

ishanaj committed:
Merge pull request #3 from sympy/master

update

ishanaj committed:
Merge pull request #4 from sympy/master

update

ishanaj committed:
solve_trig now converts hyperbolic-trig to exp

Yathartha22 commented:
It would be great if you could add more tests so as to see the flexibility of the changes.

smichr commented:
> great if you could add more tests

there are many given in #11188 and #9606...it would also be nice to discuss differences between what is done here and the advantages over those PRs.

ishanaj committed:
improved logic and added tests

_solve_trig1 first checks wether the eq is hyperbolic and then proceeds
further.

ishanaj committed:
removed test

ishanaj commented:
@smichr I was looking into the PR's. Four types of fixes have been done.
**First**, `_solve_trig` converts both types of eq(trig and hyperbolic) to `exp` and sends it to solveset for the solutions. PR #9750 , #10552.

**Second**, making seperate functions for `trig` and `hyperbolic`, with the current logic. PR #10722

**Third**, is the fix in PR #10727, which first tries a `exp(I*symbol)` substitution for trig and then tries a `exp(symbol)` subs for hyperbolic acc to the condition `if g.has(symbol) or h.has(symbol)`.

**Fourth**, using `_osborne` to convert hyperbolic to simple trig. PR #10733.

The fourth fix(`using _osborne`) seems good, but can't get why the fix has been done in `_invert_complex`

ishanaj commented:
By this commit :
```
>>>ans = solveset(sinh(x), x)
>>>ans
Union(Intersection(ImageSet(Lambda(_n, 2*_n*I*pi), S.Integers), Complexes(S.Reals x S.Reals, False)), Intersection(Complexes(S.Reals x S.Reals, False), ImageSet(Lambda(_n, I*(2*_n*pi + pi)), S.Integers)))
>>>pprint(ans)
({2*n*I*pi | n in S.Integers} n S.Complexes) U (S.Complexes n {I*(2*n*pi + pi) | n in S.Integers})
```
Also the [XFAIL](https://github.com/sympy/sympy/blob/master/sympy/solvers/tests/test_solveset.py#L610) gives:
```
>>>ans2 = solveset(sinh(x) + sech(x), x)
>>>ans2
Union(Intersection(Complexes(S.Reals x S.Reals, False), \
ImageSet(Lambda(_n, I*(2*_n*pi + pi) + log(sqrt(-2 + sqrt(5)))), S.Integers)), \
Intersection(Complexes(S.Reals x S.Reals, False), ImageSet(Lambda(_n, 2*_n*I*pi + Mod(log(sqrt(-2 + sqrt(5))), 2*I*pi)), S.Integers)), \
Intersection(Complexes(S.Reals x S.Reals, False), ImageSet(Lambda(_n, I*(2*_n*pi + pi/2) + log(sqrt(2 + sqrt(5)))), S.Integers)), \
Intersection(Complexes(S.Reals x S.Reals, False), ImageSet(Lambda(_n, I*(2*_n*pi - pi/2) + log(sqrt(2 + sqrt(5)))), S.Integers)))

>>>pprint(ans2)
/                                    /   ____________\                   \   /
|                                    |  /        ___ |                   |   |
\S.Complexes n {I*(2*n*pi + pi) + log\\/  -2 + \/ 5  / | n in S.Integers}/ U \


                          /   /   ____________\           \
                          |   |  /        ___ |           |
S.Complexes n {2*n*I*pi + \log\\/  -2 + \/ 5  / mod 2*I*pi/ | n in S.Integers}


\   /                                    /   ___________\                   \
|   |                 /         pi\      |  /       ___ |                   |
/ U |S.Complexes n {I*|2*n*pi + --| + log\\/  2 + \/ 5  / | n in S.Integers}|
    \                 \         2 /                                         /

  /                                    /   ___________\                   \
  |                 /         pi\      |  /       ___ |                   |
U |S.Complexes n {I*|2*n*pi - --| + log\\/  2 + \/ 5  / | n in S.Integers}|
  \                 \         2 /                                         /

```
Wolframalpha gives [this](http://www.wolframalpha.com/input/?i=solve+(sinh(x)%2Bsech(x)+,x))

Are they correct?

Also adding them to test file shows `SyntaxError`          
 for this part of the solution --->`....Complexes(S.Reals x S.Reals, False)...` . How to correct it?
 
ping @smichr

moorepants commented:
This needs release notes as per https://github.com/sympy/sympy/wiki/Writing-Release-Notes and also merge conflicts need to be fixed.

oscargus committed:
Merge branch 'master' into hyperbolic

oscargus committed:
Fixed merge artefact

gschintgen commented:
I think the changes done here in `_solve_trig1` have merit. Expanding upon https://github.com/sympy/sympy/pull/14053#issuecomment-362990940 I'd like to add that Osborn's rule is not applicable here, its transformations are different from actually rewriting hyperbolic functions as trigonometric functions.
Rewriting `sinh` as `sin` would be problematic too, since auto-eval will automatically revert the change for the resulting `sin(I*x)` and similar.

I checked out this branch locally, fixed it up a bit, added tests. I'd just like to make sure that I'm not stepping on anybody's toes (and get the git stuff right):
- @ishanaj : I suppose you're fine with me adding my commits to `ishanaj/sympy`?
- As for the git part: I did `git fetch upstream pull/14053/head:pr14053` and added my changes to it. Is my understanding correct, that now I'll have to add `ishanaj/sympy` as another remote for my local git repo and then push my changes to that branch? Or is it possible to push back to "pull/14053/head"?

ishanaj commented:
> @ishanaj : I suppose you're fine with me adding my commits to ishanaj/sympy?

Yes, I will be okay with it.

> As for the git part: I did git fetch upstream pull/14053/head:pr14053 and ad..........

I think you cannot push directly to this branch unless you have the right access. Once you add `ishanaj/sympy` as another remote and fetch the changes from it, you can rather create a separate sub-branch inside the `hyperbolic` branch and push your changes into it. 
I am not sure about the way you are fetching but this is what I would probably do: add `ishanaj/sympy` as remote ---->fetch the repo (git fetch 'remote')------> checkout `hyoperbolic` ----> create a  new branch inside it----->push the changes into `ishanaj/sympy`.
This would create a PR in my forked repo (ishanaj/sympy) which I can merge so that the changes reflect here.

gschintgen committed:
Merge branch 'master' into hyperbolic

(PR #14053 needs the fixed imageset / _set_function code, see #17459)

gschintgen committed:
Fix a wrong test for hyperbolic equations

The already present @XFAIL test for solveset_real asserted four
solutions, but only one of them was actually real (and valid).

gschintgen committed:
Minor fixes in _solve_trig1 for hyperbolic equations

- Use invert_real for hyperbolic case if domain.is_subset(Reals), for
exp(I*x) is no longer involved and it reduces/eliminates the chances to
hit issue #17471 which will introduce incorrect solutions.
- Fix the ConditionSets to use the correct domain instead of hardcoding
S.Reals.
- Remove superfluous intersections with S.Complexes.

gschintgen committed:
Add and reorganize tests for hyperbolic equations.

(No test has been removed.)

gschintgen commented:
Thanks @ishanaj, I've pushed my changes (and tests!) to your branch (members have push rights, so it actually worked).

I deliberately kept the changes local to `_solve_trig1` and rather minimal since this is a rather old PR and I wouldn't want to add further modifications all over the place.

Please review. (I added some more explanations in the individual commit messages.)

czgdp1807 commented:
@smichr This PR is passing all the tests. Please review so that it can be completed. Thanks.

oscarbenjamin committed:
Merge branch 'master' into hyperbolic

oscarbenjamin commented on sympy/solvers/solveset.py:
```diff
@@ -547,17 +547,23 @@ def _solve_trig(f, symbol, domain):
+    g, h = g.subs(cov, y), h.subs(cov, y)
     if g.has(symbol) or h.has(symbol):
-        return ConditionSet(symbol, Eq(f, 0), S.Reals)
+        return ConditionSet(symbol, Eq(f, 0), domain)
```
This line doesn't appear to be covered. How do we get here?

gschintgen replied:
It seems we don't.

As far as I can tell this check is intended to make `_solve_trig1` somewhat self-contained or robust in the sense that even when it's called with something out of its scope, e.g. `_solve_trig1(sin(x)+x, x, S.Reals)`, it aborts gracefully instead of going on in some "uncontrolled" circumstances where chances of succeeding are slim or non-existent.

Currently `_solve_trig1` is called by `_solve_trig` which itself is only called if a `_is_function_class` check comes to the conclusion that the equation is either composed of only trig functions or only hyperbolic trig functions. So as far as I can tell, the line you're pointing to is currently unreachable. Still, it may not be the worst idea to keep it around as a kind of "failsafe".

The line could use a test though. But there's a potential catch: it returns a ConditionSet based on `f` and not `f_original`. There's some logic in `_solve_trig` to choose the "best" ConditionSet if both `_solve_trig1` and `_solve_trig2` return ConditionSets, but I'm not entirely convinced of its correctness (I don't remember the specifics right now). Since this was completely unrelated to the issue at hand and a rather old PR to begin with, I left this check as is, except for the hard-coded S.Reals.

oscarbenjamin replied:
To me it makes more sense for internal functions like this to return `None` if they can't do anything useful. The `ConditionSet` can be constructed higher up.

gschintgen replied:
As far as I can tell, having an internal helper return a `ConditionSet` is not unusual in `solveset.py`. In this case it could arguably be constructed one level higher up, in `_solve_trig`, except that there's still the mentioned logic that is intended to choose the "best" / "simplest" `ConditionSet` representation if both `_solve_trig1` and `_solve_trig2` return (hopefully simplified) ConditionSets.

oscarbenjamin commented on sympy/solvers/solveset.py:
```diff
@@ -566,13 +572,16 @@ def _solve_trig1(f, symbol, domain):
+        if domain is S.Complexes:
+            return result
+        else:
+            return Intersection(result, domain)
```
is there nothing higher up to handle this? This doesn't seem like the right place to me. Many operations in solveset can lead to an intersection like this so simplifying those could be handled at the end. Maybe `Intersection` can be improved to handle intersections with `Complexes`.

gschintgen replied:
Yes, it may be a bit hackish to have this here, but I'd say it's not as bad as it may seem:

In general the various solveset helpers mostly return `FiniteSet`s for which intersection with `Complexes` is handled just fine, but for trig equations there are often (infinite) `ImageSet`s that (currently) don't mix well with intersections. In that sense this "intersection-removal" is somewhat specific for trig equations. Or at least here the spurious intersections are quite prominent and annoying, which prompted me to insert this workaround.
```
>>> solveset(sinh(x), x, Complexes)
{2⋅n⋅ⅈ⋅π | n ∊ ℤ} ∪ {ⅈ⋅(2⋅n⋅π + π) | n ∊ ℤ}
>>> _.intersect(Complexes)
(ℂ ∩ {2⋅n⋅ⅈ⋅π | n ∊ ℤ}) ∪ (ℂ ∩ {ⅈ⋅(2⋅n⋅π + π) | n ∊ ℤ})
```

Moving this intersection-removal too high up the call chain could also be problematic, since the origin of the spurious intersection is less well controlled.

I do concur though that it'd be best if the intersection in general could be improved, especially for "trivial" examples as shown above. (There's a recent issue for this, as I see now: #17996).

oscarbenjamin commented:
I'm commenting partly just out of curiosity as I don't know the `solveset` code that well myself. @gschintgen if you're happy with this then +1 from me to merge.

gschintgen commented:
> 
> 
> I'm commenting partly just out of curiosity as I don't know the `solveset` code that well myself. @gschintgen if you're happy with this then +1 from me to merge.

Thanks for reviewing&merging!

ishanaj commented:
Great Work!
Thanks @gschintgen !
